utils/templates/query.py.tpl

```diff
@@ -0,0 +1,361 @@
+            )
+        return name_or_query
+
+    # s.query = Q('filtered', query=s.query)
+    if hasattr(name_or_query, "_proxied"):
+        return cast(QProxiedProtocol[_T], name_or_query)._proxied
+
+    # "match", title="python"
+    return Query.get_dsl_class(name_or_query)(**params)
+
+
+class Query(DslBase):
+    _type_name = "query"
+    _type_shortcut = staticmethod(Q)
+    name: ClassVar[Optional[str]] = None
+
+    # Add type annotations for methods not defined in every subclass
+    __ror__: ClassVar[Callable[["Query", "Query"], "Query"]]
+    __radd__: ClassVar[Callable[["Query", "Query"], "Query"]]
+    __rand__: ClassVar[Callable[["Query", "Query"], "Query"]]
+
+    def __add__(self, other: "Query") -> "Query":
+        # make sure we give queries that know how to combine themselves
+        # preference
+        if hasattr(other, "__radd__"):
+            return other.__radd__(self)
+        return Bool(must=[self, other])
+
+    def __invert__(self) -> "Query":
+        return Bool(must_not=[self])
+
+    def __or__(self, other: "Query") -> "Query":
+        # make sure we give queries that know how to combine themselves
+        # preference
+        if hasattr(other, "__ror__"):
+            return other.__ror__(self)
+        return Bool(should=[self, other])
+
+    def __and__(self, other: "Query") -> "Query":
+        # make sure we give queries that know how to combine themselves
+        # preference
+        if hasattr(other, "__rand__"):
+            return other.__rand__(self)
+        return Bool(must=[self, other])
+
+
+{% for k in classes %}
+class {{ k.name }}({{ parent }}):
+    """
+    {% for line in k.docstring %}
+    {{ line }}
+    {% endfor %}
+    {% if k.kwargs %}
+
+    {% for kwarg in k.kwargs %}
+    {% for line in kwarg.doc %}
+    {{ line }}
+    {% endfor %}
+    {% endfor %}
+    {% endif %}
+    """
+    name = "{{ k.schema.name }}"
+    {% if k.params %}
+    _param_defs = {
+        {% for param in k.params %}
+        "{{ param.name }}": {{ param.param }},
+        {% endfor %}
+        {% if k.name == "FunctionScore" %}
+        "filter": {"type": "query"},
+        "functions": {"type": "score_function", "multi": True},
+        {% endif %}
+    }
+    {% endif %}
+
+    def __init__(
+        self,
+        {% if k.kwargs and not k.has_field and not k.has_fields %}
+        *,
+        {% endif %}
+        {% for kwarg in k.kwargs %}
+        {{ kwarg.name }}: {{ kwarg.type }} = NOT_SET,
+        {% endfor %}
+        **kwargs: Any
+    ):
+        {% if k.name == "FunctionScore" %}
+        if isinstance(functions, NotSet):
+            functions = []
+            for name in ScoreFunction._classes:
+                if name in kwargs:
+                    functions.append({name: kwargs.pop(name)})  # type: ignore
+        {% elif k.has_field %}
+        if not isinstance(_field, NotSet):
+            kwargs[str(_field)] = _value
+        {% elif k.has_fields %}
```

Comment:
The `has_fields` flag is for classes that are dictionaries with fields as keys. The constructor takes a single positional argument that is a dictionary.